Post to slack workflow

martin-re opened the pull request:
Implementation:
  - Workflows:
    - invoke simple HTTP request
    - post messages to SLACK
  - Workers:
    - simple HTTP task, used in mentioned workflows
  - Services:
    - HTTP service, helper service used by worker

Commit:
workflow, worker, main added

Commit:
http_service, http_worker, http-workflows added
- main to register post_to_slack
- refactor http_task (uncomplete)
- refactor post_to_slack workflow (uncomplete)
- perhaps needs to create HTTPTask for this

Commit:
pre-commit
- fixing the remaining ruff errors, would cause logic errors with env and pydantic
- (uncomplete)

Commit:
fix uri in post_to_slack wf

Commit:
fix ruff errors
- remove main, currently have not use case
- validator in HTTPInput as classmethod

Commit:
rebase main, mypy fix

Jozefiel commented on frinx/workflows/http_workflows/post_to_slack.py:
```diff
@@ -0,0 +1,33 @@
+    workflow_status_listener_enabled: bool = False
+
+    def workflow_builder(self, workflow_inputs: HTTPWorkflowService.GenericRequest.WorkflowInput) -> None:
+    #|-----------------------------------------------------------------------------------|
```
remove  it

Jozefiel commented on frinx/workflows/http_workflows/post_to_slack.py:
```diff
@@ -0,0 +1,33 @@
+
+
+class PostToSlack(HTTPWorkflowService.GenericRequest):
+#||=====================================================================================||
```
remove  it

Jozefiel commented on frinx/workers/http_workers/http_workers.py:
```diff
@@ -0,0 +1,53 @@
+
+
+class Http(ServiceWorkersImpl):
+    ###############################################################################
```
remove this comment line

Jozefiel commented on frinx/services/http_service/http_service.py:
```diff
@@ -0,0 +1,78 @@
+               json.dumps(json.loads(data)) if isinstance(data, str) else json.dumps(data)
+            except JSONDecodeError as err:
+                raise InvalidJSONError(err.args)
+        return data
```
we can move that validation to the worker level

Jozefiel commented on frinx/services/http_service/http_service.py:
```diff
@@ -0,0 +1,78 @@
+    data: dict[str, Any]
+    errors: list[str]
+    logs: Optional[list[str]] | Optional[str] | None = None
+    url: Optional[str] | None = None
```
same as input

Commit:
review comments fixes
- removing unnecessary comments
- move validation to the worker level
other modifications
- refactor post_to_slack wf
- refactor http_service.http_task
- validation of body by contentType (uncomplete)
- modify WorkflowImpl.Config.use_enum_values=True

Commit:
fix mypy error

Jozefiel commented on frinx/common/conductor_enums.py:
```diff
@@ -36,3 +36,25 @@ class RetryLogic(str, Enum):
     LINEAR_BACKOFF = 'LINEAR_BACKOFF'
+
+
+class HttpMethod(str, Enum):
+    # CONNECT = 'CONNECT'
+    # OPTIONS = 'OPTIONS'
+    # TRACE = 'TRACE'
+    DELETE = 'DELETE'
+    PATCH = 'PATCH'
+    HEAD = 'HEAD'
+    POST = 'POST'
+    PUT = 'PUT'
+    GET = 'GET'
+
+
+class ContentType(str, Enum):
+    APPLICATION_JSON = 'application/json'
+    APPLICATION_XML = 'application/xml'
+    APPLICATION_X_WWW_FORM_URLENCODED = 'application/x-www-form-urlencoded'
+    MULTIPART_FROM_DATA = 'multipart/form-data'
+    TEXT_CSV = 'text/csv'
+    TEXT_PLAIN = 'text/plain'
+    TEXT_XML = 'text/xml'
```
these ENUMS are not directly related to the conductor and should not be located here

Jozefiel replied:
maybe to services/http_service/enums.py

Jozefiel commented on frinx/workflows/http_workflows/post_to_slack.py:
```diff
@@ -0,0 +1,55 @@
+SLACK_WEBHOOK_ID_DEFAULT = 'T05ECRXU1B7/B05FA6EUAKA/8EPw07KLIxOUAhojfg2Fx7DH'
+MESSAGE_TEXT_DEFAULT = 'Hello Slack!'
+
+Worker = HTTPWorkersService.GenericHTTPWorker
```
use HTTPWorkersService.GenericHTTPWorker directly in code
this approach is less readable

martin-re replied:
Ok, I'll make a changes. Thank you.

Jozefiel commented on frinx/workflows/http_workflows/post_to_slack.py:
```diff
@@ -0,0 +1,55 @@
+
+        self.tasks.append(
+            SimpleTask(
+                task_reference_name=Worker().task_def.name,
```
not needed to construct object
use: 
name=HTTPWorkersService.GenericHTTPWorker

martin-re replied:
I need to construct object, name is instance attribute.

Jozefiel replied:
i guess, object is created in the simpleTask class

chesk how is that done in generic workflow

Jozefiel commented on frinx/workflows/http_workflows/post_to_slack.py:
```diff
@@ -0,0 +1,55 @@
+        self.tasks.append(
+            SimpleTask(
+                task_reference_name=Worker().task_def.name,
+                input_parameters=worker_input,
```
use: 
SimpleTaskInputParameters()

that object provide ability to validate compatibility during the registration

martin-re replied:
Yes, I used it.

Jozefiel commented on frinx/workflows/http_workflows/generic.py:
```diff
@@ -0,0 +1,83 @@
+                SimpleTask(
+                    name=HTTPWorkersService.GenericHTTPWorker,
+                    task_reference_name='http_task',
+                    input_parameters=SimpleTaskInputParameters(http_request=http_request),
```
is that ok? why you sending http_request object to the worker?

martin-re replied:
Yes, it's old solution. I'll make a changes.

Jozefiel commented on frinx/workflows/http_workflows/generic.py:
```diff
@@ -0,0 +1,83 @@
+
+            body: WorkflowInputField = WorkflowInputField(
+                name='body',
+                frontend_default_value={},
```
is that correctly rendered in UI? I mean is not a bad idea, but not sure if that works correctly on frontend

martin-re replied:
Good point, I'll inspect it in more detail. Thank you.

Jozefiel commented on frinx/workflows/http_workflows/generic.py:
```diff
@@ -0,0 +1,83 @@
+
+            headers: WorkflowInputField = WorkflowInputField(
+                name='headers',
+                frontend_default_value={},
```
is that correctly rendered in UI? I mean is not a bad idea, but not sure if that works correctly on frontend

martin-re replied:
Good point, I'll inspect it in more detail. Thank you.

Jozefiel commented on frinx/common/conductor_enums.py:
```diff
@@ -36,3 +36,25 @@ class RetryLogic(str, Enum):
+
+
+class HttpMethod(str, Enum):
+    # CONNECT = 'CONNECT'
+    # OPTIONS = 'OPTIONS'
+    # TRACE = 'TRACE'
```
why are these options commented?

Jozefiel replied:
and also. not explicitly conductor related

martin-re replied:
I forget to uncomment it, sorry. I'll fix it.

Jozefiel commented on frinx/services/http_service/http_service.py:
```diff
@@ -0,0 +1,77 @@
+                if not is_xml_valid(http_input['body']):
+                    raise IncompatibleFormError()
+
+    if http_input.get('accept', None):
+        kwargs['headers']['Accept'] = http_input['accept']
```
can you explain idea? why we need specific parameter for that header in http_input

martin-re replied:
Yes, this should be auth, sorry. I'll fix it.

Jozefiel commented on frinx/services/http_service/http_service.py:
```diff
@@ -0,0 +1,77 @@
+        case int(), None:
+            kwargs['timeout'] = http_input['connectTimeout']
+
+    if http_input.get('contentType', None):
+        kwargs['headers']['Content-Type'] = http_input['contentType']
+        # TODO: complete body validation by contentType
+        match http_input['contentType']:
+            case ContentType.APPLICATION_JSON:
+                if not is_json_valid(http_input['body']):
+                    raise IncompatibleFormError()
+            case ContentType.APPLICATION_XML:
+                if not is_xml_valid(http_input['body']):
+                    raise IncompatibleFormError()
+            case ContentType.APPLICATION_X_WWW_FORM_URLENCODED:
+                ...
+            case ContentType.MULTIPART_FROM_DATA:
+                ...
+            case ContentType.TEXT_CSV:
+                ...
+            case ContentType.TEXT_PLAIN:
+                ...
+            case ContentType.TEXT_XML:
+                if not is_xml_valid(http_input['body']):
+                    raise IncompatibleFormError()
```
not sure if we want to handle that complex logic.
i think, we discuss before that we don't want to handle that and definitely not in services part.

Commit:
review comments fixes
- move enums to frinx/services/http_service
- use directly reference to worker in post_to_slack
- use raw string in task_reference_name instead of workers task_def.name
- fix SimpleTaskInputParameters inputs in GenericRequest
- fix bad rendered values in GenericRequest
- uncoment commented methods in HTTPMethod
- replace accept param with basic_auth in WorkerInput
- remove body validation by Content-Type header in http_task
additional changes
- possibility to using pydantic models implicitly in worker inputs

Jozefiel commented on frinx/workers/http_workers/http_workers.py:
```diff
@@ -0,0 +1,83 @@
+                status=TaskResultStatus.COMPLETED if response.ok else TaskResultStatus.FAILED,
+                output=self.WorkerOutput(
+                    status_code=response.status_code,
+                    response=response.text,
```
Is text suitable in any situation? 
I mean if the response is not a text but JSON or any different type.

martin-re replied:
OK, I complete this with function witch can returns response also in JSON.

Jozefiel commented on frinx/workers/http_workers/http_workers.py:
```diff
@@ -0,0 +1,83 @@
+    class GenericHTTPWorker(WorkerImpl):
+
+        class WorkerDefinition(TaskDefinition):
+            name: str = 'generic_HTTP_worker'
```
use values from previous task definition             
"name": "HTTP_task",
"description": '{"description": "Generic http task", "labels": ["BASICS","HTTP"]}',

martin-re replied:
OK, it's fixed.

Commit:
Review comments fixes:
- rename worker name and description
- add possibility to returns response also in JSON
Other changes:
- use ignore comments for mypy instead of cookie_jar_to_dict function
